contracts/cw1155-base/src/contract.rs

```diff
@@ -0,0 +1,1283 @@
+        let tokens = (0..10).map(|i| format!("token{}", i)).collect::<Vec<_>>();
+        let users = (0..10)
+            .map(|i| HumanAddr::from(format!("user{}", i)))
+            .collect::<Vec<_>>();
+        let minter: HumanAddr = "minter".into();
+
+        let mut deps = mock_dependencies(&[]);
+        let msg = InitMsg {
+            minter: minter.clone(),
+        };
+        let res = instantiate(deps.as_mut(), mock_env(), mock_info("operator", &[]), msg).unwrap();
+        assert_eq!(0, res.messages.len());
+
+        execute(
+            deps.as_mut(),
+            mock_env(),
+            mock_info(minter.clone(), &[]),
+            Cw1155HandleMsg::BatchMint {
+                to: users[0].clone(),
+                batch: tokens
+                    .iter()
+                    .map(|token_id| (token_id.clone(), 1u64.into()))
+                    .collect::<Vec<_>>(),
+                msg: None,
+            },
+        )
+        .unwrap();
+
+        assert_eq!(
+            query(
+                deps.as_ref(),
+                mock_env(),
+                Cw1155QueryMsg::Tokens {
+                    owner: users[0].clone(),
+                    start_after: None,
+                    limit: Some(5),
+                },
+            ),
+            to_binary(&TokensResponse {
+                tokens: tokens[..5].to_owned()
+            })
+        );
+
+        assert_eq!(
+            query(
+                deps.as_ref(),
+                mock_env(),
+                Cw1155QueryMsg::Tokens {
+                    owner: users[0].clone(),
+                    start_after: Some("token5".to_owned()),
+                    limit: Some(5),
+                },
+            ),
+            to_binary(&TokensResponse {
+                tokens: tokens[6..].to_owned()
+            })
+        );
+
+        assert_eq!(
+            query(
+                deps.as_ref(),
+                mock_env(),
+                Cw1155QueryMsg::AllTokens {
+                    start_after: Some("token5".to_owned()),
+                    limit: Some(5),
+                },
+            ),
+            to_binary(&TokensResponse {
+                tokens: tokens[6..].to_owned()
+            })
+        );
+
+        assert_eq!(
+            query(
+                deps.as_ref(),
+                mock_env(),
+                Cw1155QueryMsg::TokenInfo {
+                    token_id: "token5".to_owned()
+                },
+            ),
+            to_binary(&TokenInfoResponse { url: "".to_owned() })
+        );
+
+        for user in users[1..].iter() {
+            execute(
+                deps.as_mut(),
+                mock_env(),
+                mock_info(users[0].clone(), &[]),
+                Cw1155HandleMsg::ApproveAll {
+                    operator: user.clone(),
+                    expires: None,
+                },
+            )
+            .unwrap();
+        }
+
+        // // FIXME test failed, query result out of order
```

Comment:
They are not "out of order", it is that the keys are stored in `CanonicalAddr`, which you cannot assume is the same ordering as `HumanAddr`. For bech32 the order sometimes is the same, sometimes different. For tests, we semi-randomize the mapping to ensure you do not depend on it.

If you hardcode some addresses, just figure out via a few test runs what the ordering is as CanonicalAddr and then assert that one. It will be stable, but it makes you be sure if you are comparing canonical or human addresses.

Discussion here: https://github.com/CosmWasm/cosmwasm/issues/552

Reply:
thanks for clarification, FIXME is resolved now.